Tests/StoreKitUnitTests/StoreMessagesHelperTests.swift

```diff
@@ -0,0 +1,164 @@
+
+        try await waitForDeferredMessages(messages: [message1, message2])
+
+        await self.helper.showStoreMessages(types: Set(StoreMessageType.allCases))
+
+        expect(message1.displayCalled) == true
+        expect(message2.displayCalled) == true
+    }
+
+    @available(iOS 16.4, *)
+    func testShowMessagesAfterDeferMessagesAndNotShowingMessagesAutomaticallyShowsSpecifiedMessages() async throws {
+        try AvailabilityChecks.iOS16_4APIAvailableOrSkipTest()
+        self.createHelper(showStoreMessagesAutomatically: false)
+
+        let message1 = MockStoreMessage(reason: Message.Reason.billingIssue)
+        let message2 = MockStoreMessage(reason: Message.Reason.priceIncreaseConsent)
+
+        try await waitForDeferredMessages(messages: [message1, message2])
+
+        await self.helper.showStoreMessages(types: [StoreMessageType.billingIssue])
+
+        expect(message1.displayCalled) == true
+        expect(message2.displayCalled) == false
+    }
+
+    @available(iOS 16.4, *)
+    func testShowMessagesAfterDeferMessagesAndShowingMessagesAutomaticallyDoesNotShowsMessages() async throws {
+        try AvailabilityChecks.iOS16_4APIAvailableOrSkipTest()
+        self.createHelper(showStoreMessagesAutomatically: true)
+
+        let message1 = MockStoreMessage(reason: Message.Reason.billingIssue)
+        let message2 = MockStoreMessage(reason: Message.Reason.priceIncreaseConsent)
+
+        try await waitForDeferredMessages(messages: [message1, message2])
+
+        await self.helper.showStoreMessages(types: Set(StoreMessageType.allCases))
+
+        expect(message1.displayCalled) == false
+        expect(message2.displayCalled) == false
+    }
+
+    @available(iOS 16.4, *)
+    func testShowMessagesAfterDeferMessagesAndNoMessagesDoesNothing() async throws {
+        try AvailabilityChecks.iOS16_4APIAvailableOrSkipTest()
+        self.createHelper(showStoreMessagesAutomatically: true)
+
+        try await waitForDeferredMessages(messages: [])
+
+        await self.helper.showStoreMessages(types: Set(StoreMessageType.allCases))
+    }
+
+    @available(iOS 16.4, *)
+    private func createHelper(showStoreMessagesAutomatically: Bool) {
+        self.helper = StoreMessagesHelper(systemInfo: self.systemInfo,
+                                          showStoreMessagesAutomatically: showStoreMessagesAutomatically,
+                                          storeMessagesProvider: self.storeMessagesProvider)
+    }
+
+    @available(iOS 16.4, *)
+    private func waitForDeferredMessages(messages: [StoreMessage]) async throws {
+        try await self.helper.deferMessagesIfNeeded()
+
+        try await Task.sleep(nanoseconds: DispatchTimeInterval.milliseconds(50).nanoseconds)
+
+        for message in messages {
+            self.storeMessagesProvider.updatesPublisher.send(message)
+
+            try await Task.sleep(nanoseconds: DispatchTimeInterval.milliseconds(50).nanoseconds)
```

Comment:
I really didn't like this... But without this, and since caching the messages is asynchronous, it would cause us to miss the messages, failing the tests... Any other suggestions are welcome

Reply:
We maaay be able to do something with a MockAsyncSequence with static values but I don't have a lot of experience with it and don't know how it would work

https://github.com/RevenueCat/purchases-ios/blob/main/Tests/UnitTests/Mocks/MockAsyncSequence.swift

Reply:
I tried but couldn't get it to work :(

Reply:
@NachoSoto thoughts?

Reply:
Can you explain why you needed to wait?

Reply:
Since we were changing threads/tasks to a bakckground task and then back to the main actor, if I didn't sleep here, the test ended before it had the chance to move the receive the message. I think that's fixed with your PR though